modules/new_serverpod_auth/serverpod_auth_session/serverpod_auth_session_flutter/lib/src/session_manager.dart

```diff
@@ -0,0 +1,155 @@
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:flutter/foundation.dart';
+import 'package:meta/meta.dart';
+import 'package:serverpod_auth_session_client/serverpod_auth_session_client.dart';
+import 'package:serverpod_auth_session_flutter/serverpod_auth_session_flutter.dart';
+import 'package:serverpod_auth_session_flutter/src/shared_preferences_key_value_storage.dart';
+
+/// Session manager client a server using `serverpod_auth_session`.
+class SessionManager extends AuthenticationKeyManager {
+  /// Creates a new session manager instance.
+  SessionManager({
+    KeyValueStorage? storage,
+
+    /// Optional distinct secure storage for the session key (e.g. Keychain).
+    KeyValueStorage? secureStorage,
+  }) : _storage = storage ?? SharedPreferencesKeyValueStorage() {
+    _secureStorage = secureStorage ?? _storage;
+  }
+
+  final KeyValueStorage _storage;
+
+  late final KeyValueStorage _secureStorage;
+
+  final _authInfo = ValueNotifier<AuthInfo?>(null);
+
+  /// The currently logged in user.
+  ValueListenable<AuthInfo?> get authInfo => _authInfo;
+
+  String? _key;
+
+  var _init = false;
+
+  /// Read the configuration from the storage
+  Future<void> init() async {
+    if (_init) {
+      throw RepeatedSessionManagerInitError();
+    }
+    _init = true;
```

Comment:
Another approach is to make the `init` function private, ie. `_init` and have `init` be a `Future<void>` instance, that is set in ctor `init = _init()`. Then you await on init multiple times, but be sure we only ran the init code once, and hence avoid this if-bad-then-exception.

Reply:
You can even hide `init` from the public interface and just have `setLoggedIn` and `logout` await it internally.